README.md
# NCKU_DL

成大製造所 110(下) 深度學習網路基礎: 理論與產業實務

---

## Description

此課程目標為學習深度學習模型與數學理論基礎，並透過Deep Learing預測股票價格作為期末專案。

* 下單交易開放時間      ：`16:00 ~ 隔日08:59`
* 台股盤中(不開放交易)  ：`09:00 ~ 14:59`

---

## Stock Price

### 獲取股票資訊

課程本身有提供基本的成交價、開盤價、收盤價等。

也可以自己抓資料，目前(20220322)是打算使用`twstock`這個別人寫好的套件來抓。

---

## Models

### 1. Prophet 預言家

為Facebook在2017所開發之模型，Prophet可以預測一些週期性的時間序列資料。

主要概念是透過多則Fourier Series來Fit時間序列：

    f(t) = year(t) + season(t) + week(t) + trend(t)

相較於自行訓練時間序列預測模型，Prophet 的一些優點如下：‌
1. 改善模型選擇和調參的時間成本：時間序列有許多經典算法如 AR, VAR, ARMA, ARIMA, 指數平滑法等，選擇模型和調參的過程可被自動化。
2. 提供讓分析師、領域專家能根據經驗法則設定的參數：例如歷史週期、特殊節日的日期等，不會因為寫成制式套件就失去自己手刻的好處。


---
## 交易API

```python
import requests

data = {
    'uname'     :<使用者名稱>,
    'pass'      :<使用者密碼>,
    'scode'     :<股票代號>,
    'svol'      :<買/賣張數>,
    'sprice'    :<買進價格>,
    'sell_price':<賣出價格>
}

r = requests.post("http://140.116.86.241:8800/api/buy", data=data)
```

stocktool/CHANGELOG.md
# Changelog

## TODO

- [ ] 技術分析指標
- [ ] Return data

## [0.1.0] - 2022-04-28
### Added
- stock file

## Reference

[CHANGELOG](https://keepachangelog.com/zh-TW/0.3.0/)

stocktool/test.py
import stock as stock

response = stock.getStock(2030)

# print(response)

stocktool2/test.py
import stock as stock
import json

response = stock.getStock(2303, period='D')

# print(response)
# print(json.dumps(response, indent=4, sort_keys=True))
with open("out.json", 'w', encoding='utf8') as f:
    json.dump(response, f, indent=4, sort_keys=False, ensure_ascii=False)

HW5.py
'''
    Using python script to simulate & calculate HW5
'''
import random

pr_get5head = 1/32
# You can customize simulation rounds for more precise
rounds = 50

# Run Experience
AtLeast1Get5Head_Round = 0
for _ in range(rounds):
    times = 0
    for i in range(100):
        p = random.uniform(0, 1)
        if(p<pr_get5head):
            times += 1
        else:
            pass
    if times > 0 :
        print(f"{times} of 100 people filp coin 5 times in a row and getting 5 head")
        AtLeast1Get5Head_Round+=1
    else:
        print("There is no one")
print(f"Run Simulation {rounds} times, and we can get prob:{AtLeast1Get5Head_Round/rounds}")
print()
# Calculate answer by math
Prob = 1 - ( ( 1-(1/32) )**100 )
print(f"The prob. of [At least 1 person get 5 head] calculated by math  1-( (1-(1/32))**100) is {Prob}")


model.py
import json
import prophet
import pandas as pd
from twstock import *

print(prophet.__version__)
# 印出所有台股證券編碼
# print(json.dumps(twstock.codes, indent=4, ensure_ascii=False))

stock_name = {
    '2330': u'台積電',
    '2317': u'鴻海',
    '1301': u'台塑',
    '1326': u'台化',
    '2412': u'中華電',
    '3008': u'大立光',
    '1303': u'南亞',
    '2308': u'台達電',
    '2454': u'聯發科',
    '2881': u'富邦金',
    '8299': u'群聯',
    '6223': u'旺矽'
}

stock_list = [
    '2330',  # 台積電
    '2317',  # 鴻海
    '1301',  # 台塑
    '1326',  # 台化
    '2412',  # 中華電
    '3008',  # 大立光
    '1303',  # 南亞
    '2308',  # 台達電
    '2454',  # 聯發科
    '2881',  # 富邦金
    '8299',  # 群聯
    '6223',  # 旺矽
]



stock = Stock('2330')
# 印出近31日 [2330] 收盤價 
print(stock.price)
# # 印出近31日 [2330] 最高價 
# print(stock.high)
# # 印出近31日 [2330] 成交量
# print(stock.transaction)

'''
    Prophet單變量時間序列預測
'''
df = pd.array(stock.price)
print(df)

classstock.py
import json
import requests
from datetime import datetime, timedelta

class Stock:
    data = []
    def __init__(
            self, 
            stockNo:int, 
            days:int = 30, 
            start:str = (datetime.today() - timedelta(days=30)).strftime("%Y%m%d"), 
            end:str = datetime.today().strftime("%Y%m%d")
        ):
        '''
        Parameters
        ----------
        stockNo   : int (required) 
            個股代號
        days      : int (optional) 
            日期區間
        start     : str (optional) 
            查詢區間的起始時間, 預設為30天前, 必須是YYYYMMDD格式
        end       : str (optional) 
            查詢區間的結束時間, 預設為當天日期, 必須是YYYYMMDD格式
        
        '''
        self.stockNo = stockNo
        self.start_time = start
        self.end_time = end
        print(self.stockNo, self.start_time, self.end_time)

        print(json.dumps(self.getStockInfo(), indent=4, sort_keys=True))
        # print(self.getStockInfo())

    # 從MMDB獲取個股資訊
    def getStockInfo(self):
        information_url = ('http://140.116.86.242:8081/stock/' +
                            'api/v1/api_get_stock_info_from_date_json/' +
                            str(self.stockNo) + '/' +
                            str(self.start_time) + '/' +
                            str(self.end_time)
                            )
        result = requests.get(information_url).json()
        if(result['result'] == 'success'):
            return result['data']
        return dict([])

    # 計算移動平均線
    def getMA(self):
        pass

    # 計算KD

    # 計算MACD

    # 計算RSI

    # 計算乖離率

    # 計算威廉指標

    # 多空指標乖離

    # CDP

    # DMI

    

def main():
    stock=Stock(2330)

if __name__ == "__main__":
    main()

stocktool1/tool1.py
import json
import requests
import time
from bs4 import BeautifulSoup
import requests

def layer2crawler(id):
    print(id)
    base_url = "https://www.wantgoo.com/index/"+"%5E"+str(id)+"/stocks"
    agent = {"User-Agent": 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/100.0.4896.127 Safari/537.36 Edg/100.0.1185.50'}
    r = requests.get(base_url, headers=agent)
    # print(r.content.decode('unicode-escape'))
    # contents = json.loads(r.content.decode('unicode-escape'))
    soup = BeautifulSoup(r.text, 'html.parser')
    cls = soup.find_all('tr')
    for c in cls:
        print(c)
    return "test"

def webCrawler():
    base_url = "https://www.wantgoo.com/index/all-industry-of-groups"
    agent = {"User-Agent": 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/100.0.4896.127 Safari/537.36 Edg/100.0.1185.50'}
    r = requests.get(base_url, headers=agent)
    # print(r.content.decode('unicode-escape'))
    contents = json.loads(r.content.decode('unicode-escape'))
    soup = BeautifulSoup(r.text, 'html.parser')
    data = dict()
    idListA = []
    idListB = []
    for c in contents:
        groupId:str = ""
        if len(str(c['id']))==2:
            groupId = "^0"+str(c['id'])
        else:
            groupId = "^"+str(c['id'])
        if c['groupName'] == '上市':
            idListA.append(groupId)
            # groupA.append({
            #     "id": groupId,
            #     "groupName": c['shortName'] 
            # })
        elif c['groupName'] == '上櫃':
            idListB.append(groupId)
            # groupB.append({
            #     "id": groupId ,
            #     "groupName": c['shortName']
            # })
    data['上市'] = idListA
    data['上櫃'] = idListB
    print(data['上市'])
    return data
    # print(layer2crawler(20))
    # print(json.dumps(data, indent=4, sort_keys=True, ensure_ascii=False))
    
def allalive(idList):
    base_url = "https://www.wantgoo.com/investrue/all-alive"
    agent = {"User-Agent": 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/100.0.4896.127 Safari/537.36 Edg/100.0.1185.50'}
    r = requests.get(base_url, headers=agent)
    # print(r.content.decode('unicode-escape'))
    contents = json.loads(r.content.decode('unicode-escape'))
    # print(contents[399:402])
    soup = BeautifulSoup(r.text, 'html.parser')
    data = []
    dictList = dict()
    searchDict = dict()
    for c in contents:
        searchDict[c['id']] = c['name']
        if len(c['industries']) < 1:
            continue
        for i in range(len(c['industries'])):
            if c['industries'][i]['id'] in idList["上市"]:
                # print(c['name'], c['id'])
                # print(c['industries'][i]['id'], c['industries'][i]['name'], c['industries'][i]['shortName'])
                if c['industries'][i]['name'] not in dictList.keys():
                    dictList[c['industries'][i]['name']] = []
                dictList[c['industries'][i]['name']].append({
                    "id":c['id'], 
                    "name":c['name']
                })
            if c['industries'][i]['id'] in idList["上櫃"]:
                if c['industries'][i]['name'] not in dictList.keys():
                    dictList[c['industries'][i]['name']] = []
                dictList[c['industries'][i]['name']].append({
                    "id":c['id'], 
                    "name":c['name']
                })
    return dictList, searchDict
    # print(dictList)
if __name__ == "__main__":
    data = webCrawler()
    outputData, searchList = allalive(data)
    print(len(outputData), outputData.keys())
    with open('stock.json', 'w', encoding='utf8') as outfile:
        json.dump(outputData, outfile, indent=4, ensure_ascii=False)
    with open('search.json', 'w', encoding='utf8') as out:
        json.dump(searchList, out, indent=4, ensure_ascii=False)
    # print(json.dumps(outputData, indent=4, sort_keys=True, ensure_ascii=False))  


Random_Buy_Example.py
import random
import requests
from datetime import datetime
import time

# 取得股票資訊
# Input:
#   stock_code: 股票ID
#   start_date: 開始日期，YYYYMMDD
#   stop_date: 結束日期，YYYYMMDD
# Output: 股票資訊陣列
'''
    Example output:
    {   
        'capacity': 1214965,
        'change': 0.5,
        'close': 136,
        'date': 1649376000,
        'high': 137.5,
        'low': 135.5,
        'open': 136.5,
        'shdid': 23948326,
        'stock_code_id': '2492',
        'transaction_volume': 1189,
        'turnover': 165453618
    }
'''


def Get_Stock_Informations(stock_code, start_date, stop_date):
    information_url = ('http://140.116.86.242:8081/stock/' +
                       'api/v1/api_get_stock_info_from_date_json/' +
                       str(stock_code) + '/' +
                       str(start_date) + '/' +
                       str(stop_date)
                       )
    result = requests.get(information_url).json()
    if(result['result'] == 'success'):
        return result['data']
    return dict([])

# 取得持有股票
# Input:
#   account: 使用者帳號
#   password: 使用者密碼
# Output: 持有股票陣列


def Get_User_Stocks(account, password):
    data = {'account': account,
            'password': password
            }
    search_url = 'http://140.116.86.242:8081/stock/api/v1/get_user_stocks'
    result = requests.post(search_url, data=data).json()
    if(result['result'] == 'success'):
        return result['data']
    return dict([])

# 預約購入股票
# Input:
#   account: 使用者帳號
#   password: 使用者密碼
#   stock_code: 股票ID
#   stock_shares: 購入張數
#   stock_price: 購入價格
# Output: 是否成功預約購入(True/False)


def Buy_Stock(account, password, stock_code, stock_shares, stock_price):
    print('Buying stock...')
    data = {'account': account,
            'password': password,
            'stock_code': stock_code,
            'stock_shares': stock_shares,
            'stock_price': stock_price}
    buy_url = 'http://140.116.86.242:8081/stock/api/v1/buy'
    result = requests.post(buy_url, data=data).json()
    print('Result: ' + result['result'] + "\nStatus: " + result['status'])
    return result['result'] == 'success'

# 預約售出股票
# Input:
#   account: 使用者帳號
#   password: 使用者密碼
#   stock_code: 股票ID
#   stock_shares: 售出張數
#   stock_price: 售出價格
# Output: 是否成功預約售出(True/False)


def Sell_Stock(account, password, stock_code, stock_shares, stock_price):
    print('Selling stock...')
    data = {'account': account,
            'password': password,
            'stock_code': stock_code,
            'stock_shares': stock_shares,
            'stock_price': stock_price}
    sell_url = 'http://140.116.86.242:8081/stock/api/v1/sell'
    result = requests.post(sell_url, data=data).json()
    print('Result: ' + result['result'] + "\nStatus: " + result['status'])
    return result['result'] == 'success'


# 隨機購入或出售
# Input: None
# Output: None
def Random_Buy_Or_Sell():

    account = '帳號'  # 使用者帳號
    password = '密碼'  # 使用者密碼

    today = datetime.today().strftime('%Y%m%d')  # 今日日期，YYYYMMDD

    action = random.randrange(0, 10000000) & 1  # 決定操作為隨機購買或售出，0=購買、1=售出
    user_stocks = Get_User_Stocks(account, password)  # 取得使用者持有股票S
    if(len(user_stocks) == 0):  # 若使用者不持有任何股票
        action = 0  # 指定操作為購買股票
    if(action == 0):  # 如果操作為購買股票
        target_stocks_id = [2330]  # 購買股票清單，隨機購買將於該清單內隨機挑選一個股票來進行購買

        selected_stock_id = target_stocks_id[random.randrange(
            0, 100000000) % len(target_stocks_id)]  # 於購買清單隨機挑選一張股票

        today_stock_information = Get_Stock_Informations(
            selected_stock_id, '20200101', today)  # 取得選定股票往日資訊
        if(len(today_stock_information) == 0):  # 若選定股票沒有任何資訊
            print('未曾開市')
            return  # 結束操作
        today_stock_information = today_stock_information[0]  # 取得選定股票最新資訊
        # 股票隨機浮動範圍設定為最高價-最低價
        random_trading_price_offset_range = today_stock_information['high'] - \
            today_stock_information['low']

        buy_shares = random.randrange(1, 10)  # 股票隨機數量選定1~10張

        #購買價格以收盤價格-浮動範圍*(隨機浮點數0.0~1.0)
        buy_price = today_stock_information['close'] - random.random() * random_trading_price_offset_range
        # buy_price = today_stock_information['high']  # 購買價格設定為股票最高價

        Buy_Stock(account, password, selected_stock_id,
                  buy_shares, buy_price)  # 購買股票
    else:  # 操作為售出股票
        selected_stock = user_stocks[random.randrange(
            0, 100000000) % len(user_stocks)]  # 隨機選定一個使用者持有的股票

        selected_stock_id = selected_stock['stock_code_id']  # 取得選定的股票ID

        today_stock_information = Get_Stock_Informations(
            selected_stock_id, '20200101', today)  # 取得選定股票最新資訊
        if(len(today_stock_information) == 0):  # 若選定股票沒有任何資訊
            print('未曾開市')
            return  # 結束操作
        today_stock_information = today_stock_information[0]  # 取得選定股票最新資訊
        # 股票隨機浮動範圍設定為最高價-最低價
        random_trading_price_offset_range = today_stock_information['high'] - \
            today_stock_information['low']

        keeping_shares = selected_stock['shares']  # 取得使用者在選定股票所持有的張數
        sell_shares = random.randrange(
            1, keeping_shares)  # 隨機取得售出張數(1~使用者持有張數)
        sell_price = today_stock_information['close'] + random.random(
        ) * random_trading_price_offset_range  # 購買價格以收盤價格+浮動範圍*(隨機浮點數0.0~1.0)

        Sell_Stock(account, password, selected_stock_id,
                   sell_shares, sell_price)  # 售出股票


Random_Buy_Or_Sell()


stocktool2/stock.py
import json
import requests
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from pandas import json_normalize
from datetime import date, datetime, timedelta

def getStockInformations(stockNo, start_date, end_date):
    information_url = ('http://140.116.86.242:8081/stock/' +
                       'api/v1/api_get_stock_info_from_date_json/' +
                       str(stockNo) + '/' +
                       str(start_date) + '/' +
                       str(end_date)
                       )
    result = requests.get(information_url).json()
    if(result['result'] == 'success'):
        return result['data']
    return dict([])

def SUM(S:pd.DataFrame, N:int):
    return  pd.Series(S).rolling(N).sum() if N>0 else pd.Series(S).cumsum()

#   對Series整體下移動 N, 返回 Series(shift後會產生NAN)
def REF(S:pd.DataFrame, N:int):
    return pd.Series(S).shift(N)  

def IF(S_BOOL,S_TRUE,S_FALSE):          
    return np.where(S_BOOL, S_TRUE, S_FALSE)

#   Series max
def MAX(S1,S2):  
    return np.maximum(S1,S2)    

def __changePeriod(stockInfo, period='W-mon'):
    df = json_normalize(stockInfo)
    df = df.sort_values(by=['date'], ascending=True)
    df.index = pd.to_datetime(df['date'], unit='s')
    rf = df.resample(period).last()
    rf['high'] = df['high'].resample(period).max()
    rf['low'] = df['low'].resample(period).min()
    rf['change'] = df['change'].resample(period).sum()
    # weekdf.fillna(method='ffill', inplace=True)

    return json.loads(rf.to_json(orient='records'))

def __calculateMA(df):
    '''計算移動平均線(MA), 將N天的收盤價加總後再除以N, 即得到第N天的算術平均線數值。
    
    常用的日線有: 5日線, 10日線, 20日線, 60日線, 120日線, 240日線。
    '''
    # 分別計算5天, 10天, 20天和60天的移動平均線
    df['MA_5'] = df['close'].rolling(5).mean()
    df['MA_10'] = df['close'].rolling(10).mean()
    df['MA_20'] = df['close'].rolling(20).mean()
    df['MA_60'] = df['close'].rolling(60).mean()
    return df
    
def __calculateKD(df):
    '''計算KD值
    
    '''

    # 計算9日內最高成交價
    df['9daymax'] = df['high'].rolling(9).max()
    df['9daymin'] = df['low'].rolling(9).min()
    # print(df['9daymax'], df['9daymin'])
    df['RSV'] = 0
    df['RSV'] = (df['close'] - df['9daymin']) / (df['9daymax'] - df['9daymin'])*100
    # print(df['RSV'])

    # 計算k值 計算d值
    df['K9'] = df['RSV'].ewm(com=2, adjust=False).mean()
    df['D9'] = df["K9"].ewm(com=2, adjust=False).mean()
    df["3K-2D"] = 3 * df["K9"] - 2 * df["D9"]
    # df.reset_index(inplace=True)
    # df['date'] = pd.to_datetime(df['date'], unit='s')
    df = df.drop(['9daymax', '9daymin'], axis=1)
    return df

def __calculateMACD(df):
    '''MACD, Moving Average Convergence and Divergence, 股市中常說的平滑異同移動平均線。由DIF、DEA、MACD組成

    MA分成EMA(exponential)和SMA(mean), 這裡使用EMA

    '''
    # com=1 代表權重為 1/(com+1) = 1/2
    df['EMA_12'] = df['close'].ewm(span=12, adjust=False).mean()
    df['EMA_26'] = df['close'].ewm(span=26, adjust=False).mean()
    # df['EMA_12'] = df['EMA_12'].rolling(12).mean()
    # df['EMA_26'] = df['EMA_26'].rolling(26).mean()
    # 差離值
    df['DIF9'] = df['EMA_12'] - df['EMA_26']
    # 差離平均值
    df['DEA'] = df['DIF9'].ewm(span=9, adjust=False).mean()
    # yahoo 以 DEA 為 MACD
    df['MACD'] = df['DEA']
    # 柱狀值
    # df['MACD'] = 2*(df['DIF'] - df['DEA'])
    return df.drop(['DEA'], axis=1)

def __calculateRSI(df):
    '''計算RSI值
    
    RSI是計算過去一段時間的相對強弱, 這個"一段時間"是看個人選擇要多長的時間, 一般比較常見的是6日、12日、14日、24日。

    一樣預設使用EMA

    '''
    close_delta = df['close'].diff()

    high = close_delta.clip(lower=0)
    low = -1*close_delta.clip(upper=0)

    period = 5
    maHigh = high.ewm(com=period-1, adjust=True, min_periods=period).mean()
    maLow = low.ewm(com=period-1, adjust=True, min_periods=period).mean()

    df['RSI5'] = maHigh/maLow
    df['RSI5'] = 100-(100/(1+df['RSI5']))

    period = 10
    maHigh = high.ewm(com=period-1, adjust=True, min_periods=period).mean()
    maLow = low.ewm(com=period-1, adjust=True, min_periods=period).mean()

    df['RSI10'] = maHigh/maLow
    df['RSI10'] = 100-(100/(1+df['RSI10']))

    return df

def __calculateBIAS(df):
    '''計算乖離率(bias), 公式為: (目前價格 - MA) / MA, 乖離率所衡量的就是現行股價與均線的差異程度。
    
    乖離率為正, 意即股價>平均線；乖離率為負, 意即股價<平均線。
    因為乖離率的計算和MA有關, 因此乖離率也會因為使用的日均線天數而有所不同。

    '''
    df['BIAS_5'] = 100*((df['close']-df['MA_5'])/df['MA_5'])
    df['BIAS_10'] = 100*((df['close']-df['MA_10'])/df['MA_10'])
    df['BIAS_20'] = 100*((df['close']-df['MA_20'])/df['MA_20'])
    df['B10-B20'] = df['BIAS_10'] - df['BIAS_20']
    return df

def __calculateWR(df):
    '''計算威廉指標
    
    以14天為計算週期

    原本的威廉指標，值越小時（收盤價越接近最高價），代表市場越處於超買中。
    
    有些會乘上(-1)使得該指標符合"指標值越大代表超買、越小則代表超賣的統一認知"。
    
    '''
    high = df['high'].rolling(9).max()
    low = df['low'].rolling(9).min()
    close = df['close']
    df['W%R9'] = 100 * ((high - close) / (high - low))
    return df

def __calculateBBI(df):
    '''計算多空指標(BBI), Bull and Bear Index, 將不同日數移動平均線加權平均之後的綜合指標, 屬於均線型指標。

    一般將3日、6日、12日和24日的四種sma(或ema)作為計算的參數。

    計算公式: BBI=(3日均價+6日均價+12日均價+24日均價)÷4

    '''
    # 分別計算3天, 6天, 12天和24天的移動平均線
    df['MA_3'] = df['close'].rolling(3).mean()
    df['MA_6'] = df['close'].rolling(6).mean()
    df['MA_12'] = df['close'].rolling(12).mean()
    df['MA_24'] = df['close'].rolling(20).mean()
    # df['MA_3'] = df['close'].ewm(span=3, adjust=False).mean()
    # df['MA_6'] = df['close'].ewm(span=6, adjust=False).mean()
    # df['MA_12'] = df['close'].ewm(span=12, adjust=False).mean()
    # df['MA_24'] = df['close'].ewm(span=20, adjust=False).mean()
    df['M3'] = df['MA_3']
    df['BS'] = (df['MA_3']+df['MA_6']+df['MA_12']+df['MA_24'])/4
    df['M3-BS'] = df['M3'] - df['BS']
    # return df
    return df.drop(['MA_3', 'MA_6', 'MA_12', 'MA_24'], axis=1)

def __calculateCDP(df):
    '''計算CDP, Contrarian Operation 
    
    計算公式: CDP = (H+L+C*2)÷4 , H: 前一日最高價, L: 前一日最低價, C: 前一日收盤價

    '''

    df['CDP'] = (df['high']+df['low']+df['close']*2)/4
    df['AN'] = df['CDP'] + (df['high'] - df['low'])
    df['NH'] = df['CDP']*2 - df['low']
    df['AL'] = df['CDP'] - (df['high'] - df['low'])
    df['NL'] = df['CDP']*2 - df['high']

    return df

def __calculateDI(df):
    '''計算DI, Directional Movement Index, 這個指標是由四個指標組合而成, 分別是ADX,ADXR,PDI,MDI。
    
    '''

    df['HL'] = df['high'] - df['low']
    df['DM+'] = (df['high'] - df['close'].shift(1)).abs()
    df['DM-'] = (df['low'] - df['close'].shift(1)).abs()
    df['TR'] = df[['HL','DM+','DM-']].max(axis=1)
    del df['HL'], df['DM+'], df['DM-']

    # df['ATR'] = df['TR'].ewm(span=14, adjust=False).mean()
    df['ATR'] = df['TR'].rolling(14).mean()

    # +-DX
    df['HD'] = df['high'] - REF(df['high'], 1)
    df['LD'] = REF(df['low'], 1) - df['low'] 
    df['+DX'] = IF((df['HD'] > 0) & (df['HD'] > df['LD']), df['HD'], 0)
    df['S+DX'] = df['+DX'].rolling(14).mean()
    # df['S+DX'] = df['+DX'].ewm(span=14, adjust=False).mean()
    df['-DX'] = IF((df['LD'] > 0) & (df['LD'] > df['HD']), df['LD'], 0)
    df['S-DX'] = df['-DX'].rolling(14).mean()
    # df['S-DX'] = df['-DX'].ewm(span=14, adjust=False).mean()

    # +-DI
    df['+DI'] = (df['S+DX'] / df['ATR'])*100         
    df['-DI'] = (df['S-DX'] / df['ATR'])*100

    # ADX
    df['ADX'] = (np.abs(df['+DI'] - df['-DI'])/(df['+DI'] + df['-DI']))*100
    # df['ADX'] = pd.Series(np.abs(df['-DI'] - df['+DI']) / (df['+DI'] + df['-DI'])*100)
    df['ADX'] = df['ADX'].rolling(14).mean()
    df['S-DX'] = df['-DX'].ewm(span=14, adjust=False).mean()
    df['ADXR'] = (df['ADX'] + REF(df['ADX'], 14)) / 2

    # return df
    return df.drop(['+DX','-DX', 'S+DX','S-DX', 'TR', 'HD', 'LD'], axis=1)

def getStock(stockNo:int, start_date:str=None, end_date:str=None, period:str='D'):
    '''
    Parameters
    ----------
    stockNo   : int (required) 
        個股代號
    start     : str (optional) 
        查詢區間的起始時間, 預設為90天前(一個季), 必須是YYYYMMDD格式
    end       : str (optional) 
        查詢區間的結束時間, 預設為當天日期, 必須是YYYYMMDD格式
    period    : str (optional)
        指定使用日線周線月線
    '''
    if start_date == None:
        start_date = (datetime.today() - timedelta(days=180)).strftime("%Y%m%d")
    if end_date == None:
        end_date = datetime.today().strftime("%Y%m%d")

    stockInfo = getStockInformations(stockNo, start_date, end_date)
    stockName:str = ""
    with open('search.json', 'r', encoding='utf8') as inf:
        slist = json.load(inf)
        stockName = slist[str(stockNo)]
    
    
    if period=='W':
        stockInfo = __changePeriod(stockInfo)
    elif period=='M':
        stockInfo = __changePeriod(stockInfo, period='M')


    df = json_normalize(stockInfo)
    df = df.sort_values(by=['date'], ascending=True)
    df.index = pd.to_datetime(df['date'], unit='s')

    df = __calculateMA(df)

    df = __calculateKD(df)
    
    df = __calculateMACD(df)

    df = __calculateRSI(df)
    
    df = __calculateBIAS(df)

    df = __calculateWR(df)

    df = __calculateBBI(df)
    
    df = __calculateCDP(df)

    df = __calculateDI(df)

    df = df.drop(['stock_code_id'], axis=1)
    
    response = dict()
    response['stockNo'] = stockNo
    response['stockName'] = stockName
    data = json.loads(df.to_json(orient='records'))
    response['start_date'] = datetime.fromtimestamp(data[0]['date']).strftime('%Y%m%d')
    response['end_date'] = datetime.fromtimestamp(data[-1]['date']).strftime('%Y%m%d')
    dataList = []
    dateList = dict()
    for d in data:
        time = datetime.fromtimestamp(d['date']).strftime('%Y%m%d')
        d.pop('date')
        dataList.append(dict({time:d}))
        dateList[time] = d
    
    response['data'] = dateList
    return response